[nebula] Add Nebula (watchnebula.com) extractor

hheimbuerger opened the pull request:
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [X] New extractor

---

### Description of your *pull request* and other information

This PR is a copy of https://github.com/ytdl-org/youtube-dl/pull/27859 (as of Feb 2021: in status Open)
However, that PR contains the full commit history (pretty clean and readable), while this one here is simply the squashed version.

This is a new extractor for the video platform Nebula (https://watchnebula.com/), created by the streamer community Standard. Nebula uses the Zype video infrastructure and this extractor is using the url_transparent mode to hand off video extraction to the Zype extractor (which has shown to be super reliable here).

All videos on Nebula require a subscription to access. Multiple authentication methods are supported, with the most convenient probably being .netrc, but a few alternatives (cookie jar, environment variable, token in --videopassword) have been implemented as well.

As there are no known freely available videos, the three test cases only pass when an auth token to a user account with a (paid) subscription is present. This can be specified using the NEBULA_TOKEN environment variable (as I couldn't figure out a way to use --netrc in unit tests).

I have aimed to provide comprehensive documentation of approach and behavior in class and method docstrings.

Initial discussion has occured on issue https://github.com/ytdl-org/youtube-dl/issues/21258. I didn't open the issue, but hijacked it for my implementation.
Afterwards, development discussion has occured over multiple months on PR https://github.com/ytdl-org/youtube-dl/pull/24805. The fork that PR was based on has been destroyed by the DMCA takedown. That's why it had to be reopened as https://github.com/ytdl-org/youtube-dl/pull/27859.

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+                self.to_screen('Authenticating to Nebula with token from NEBULA_TOKEN environment variable')
+
+        # option #4: nebula token via --videopassword
+        if not nebula_token:
+            nebula_token = self._downloader.params.get('videopassword')
+            if nebula_token:
+                self.to_screen('Authenticating to Nebula with token from --videopassword')
```
It looks to me like using `--video-password` is a misuse for this extractor. If it is not, please explain; and add this website to the documentation in `options.py`

hheimbuerger replied:
You're right, it's totally a misuse. Authentication was a bit under-documented, so in the beginning I was just trying to find out various different ways to authenticate, to find one that's rather clean. It took a good while until a reviewer or used suggested netrc, I hadn't even heard of it before (and the docs didn't really mention it).

So this just stayed around, because it had already been implemented. But I agree, I don't think anyone's ever gonna use it, so I'll remove it.

hheimbuerger replied:
Support for `--videopassword` has now been removed.

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+
+        # option #2: nebula token via cookie jar
+        if not nebula_token:
+            # TODO: is there a helper to do all this cookie extraction?
+            nebula_cookies = self._get_cookies('https://watchnebula.com')
+            nebula_cookie = nebula_cookies.get('nebula-auth')
```
no, doesnt look like there is any helper

```suggestion
            nebula_cookie = self._get_cookies('https://watchnebula.com').get('nebula-auth')

```

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+                nebula_cookie_value = compat_urllib_parse_unquote(nebula_cookie.value)
+                nebula_token = self._parse_json(nebula_cookie_value, video_id).get('apiToken')
+
+        # option #3: nebula token via environment variable
+        if not nebula_token and 'NEBULA_TOKEN' in os.environ:
+            nebula_token = os.environ.get('NEBULA_TOKEN')
+            if nebula_token:
+                self.to_screen('Authenticating to Nebula with token from NEBULA_TOKEN environment variable')
+
```
Dont use environment

hheimbuerger replied:
I currently need this to run the unit tests. Can you tell me how else to do auth (e.g. to activate netrc support) when running the tests?

pukkandan replied:
Add a `'usenetrc': True` in `params` of the tests

hheimbuerger replied:
Yep, got that working, awesome! Removing the envvar support then.

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+                'timestamp': 1533009600,
+                'channel': 'Lindsay Ellis',
+                'uploader': 'Lindsay Ellis',
+            }
```
```suggestion
            },
            'skip': 'All WatchNebula content requires authentication',
```

Same for other tests

hheimbuerger replied:
Done. ✅

Is there a way to ignore the skips for my own unittest runs? Something like a `--no-skip` command line argument? (I quickly read through the `test_template()` method, but didn't spot a way to do that.)

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+
+from .common import InfoExtractor
+from ..compat import compat_urllib_parse_unquote, compat_str
+from ..utils import parse_iso8601, ExtractorError, try_get, urljoin, sanitized_Request
```
```suggestion
from ..utils import (
    ExtractorError, 
    parse_iso8601,
    sanitized_Request,
    try_get,
    urljoin,
)
```

for readablity

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+
+
+class NebulaIE(InfoExtractor):
+    """
+    Nebula (https://watchnebula.com/) is a video platform created by the streamer community Standard. It hosts videos
+    off-YouTube from a small hand-picked group of creators.
+
+    All videos require a subscription to watch. There are no known freely available videos. An authentication token to
+    an account with a valid subscription can be specified in multiple ways, including credentials in .netrc or a cookie
+    jar.
+    As neither of these parameters appear to be supported by the unit test runner, it's recommended to set the envvar
+    NEBULA_TOKEN to execute the test runs.
+
+    Nebula uses the Zype video infrastructure and this extractor is using the 'url_transparent' mode to hand off
+    video extraction to the Zype extractor.
+
+    This description has been last updated on 2020-10-22.
+    """
```
remove

hheimbuerger replied:
Are you willing to negotiate that?

I think this class documentation gives a good overview of what the extractor is for. ('Nebula' does not seem a particularly globally unique name, even if I'm not aware of another platform with that name today.)

And then it gives a broad introduction to how the extractor works. I think that's a good stepping stone for anyone else going into this code, and should ease code understanding by a decent amount.

Of course I'll update this as the PR progresses, e.g. the removed alternate authentication methods.

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+    This description has been last updated on 2020-10-22.
+    """
+
+    _VALID_URL = r'https?://(?:www\.)?watchnebula\.com/videos/(?P<id>[-\w]+)'   # the 'id' group is actually the display_id, but we misname it 'id' to be able to use _match_id()
```
> the 'id' group is actually the display_id, but we misname it 'id' to be able to use _match_id()

This could potentially cause issues with archive. I'm not sure though and need to investigate further

hheimbuerger replied:
Can you give me a hint (e.g. deep link into docs) for what you refer to by 'archive'?

Would you prefer `_match_id()` not being used for this extractor? It's been a while, this line is a year old, but I think there was some other consequences to not calling that method at all.

pukkandan replied:
> Can you give me a hint (e.g. deep link into docs) for what you refer to by 'archive'?

I mean `--download-archive`. I didnt get time to check in-depth whether this needs to be changed. You dont need to change anything for now. This was more of a note to myself that this needs investigation b4 merging

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+        return response['key']
+
+    def _retrieve_nebula_auth(self, video_id):
+        """
+        Attempt to find a Nebula API token. Makes multiple attempts in the
+        following order:
+        a) login credentials used to authenticate to the Nebula login endpoint,
+           either from .netrc or specified using --username/--password
+        b) the --cookies supplied cookie jar
+        c) the NEBULA_TOKEN environment variable
+        d) the --video-password command line argument (this isn't documented in
+           the error message, because probably highly unpopular)
+        If none of these are successful, an end user-intended error message is
+        raised, listing some solutions.
+
+        Returns a Nebula API token, which subsequently can be used to make
+        authenticated calls to the Nebula API.
+        """
```
```suggestion
        """Returns a Nebula API token, which subsequently can be used to make authenticated calls to the Nebula API"""
```

hheimbuerger replied:
Vastly simplified the docstring (albeit not quite to the suggested version), after three out of four authentication methods have been removed.

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+        """
+        nebula_token = None
+
+        # option #1: login credentials via .netrc or --username and --password
```
there is a lot of comments explaining things that are obvious from the code. This only makes the code harder to read. Please remove these

hheimbuerger replied:
As discussed on Discord, I respectfully disagree, but I will oblige as much as possible. 😒

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+        return nebula_token
+
+    def _retrieve_zype_api_key(self, page_url, display_id):
+        """
+        Retrieves the Zype API key required to make calls to the Zype API.
+
+        Unfortunately, the Nebula frontend stores this as a JS object literal in one of its JS chunks,
+        looking somewhat like this (but minified):
+
+            return {
+                NODE_ENV: "production",
+                REACT_APP_NAME: "Nebula",
+                REACT_APP_NEBULA_API: "https://api.watchnebula.com/api/v1/",
+                REACT_APP_ZYPE_API: "https://api.zype.com/",
+                REACT_APP_ZYPE_API_KEY: "<redacted>",
+                REACT_APP_ZYPE_APP_KEY: "<redacted>",
+                // ...
+            }
+
+        So we have to find the reference to the chunk in the video page (as it is hashed and the hash will
+        change when they do a new release), then download the chunk and extract the API key from there,
+        hoping they won't rename the constant.
+
+        Alternatively, it is currently hardcoded and shared among all users. We haven't seen it
+        change so far, so we could also just hardcode it in the extractor as a fallback.
+        """
```
```suggestion
        """Retrieves the Zype API key required to make calls to the Zype API"""
```

hheimbuerger replied:
I think this is very relevant context and prior research for anyone (including myself) who has to maintain this code in the future, especially if it breaks because they make changes to their frontend. Why would you want this removed?

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+            access_token=zype_access_token)
+
+    def _extract_channel(self, video_meta):
+        """
+        Extract the channel title, by going through the list of categories and finding the first value of the
+        first category that has a value.
+
+        I know this look like a terrible approach. But actually, it's just reproducing the behavior of the
+        React code the Nebula frontend uses (as of 2020-04-07):
+
+            let channel;
+            if (video && video.categories && video.categories.length) {
+                const channelTitle = video.categories.map((category) => (category.value[0]))
+                                                     .filter((title) => (!!title))[0];
+                channel = getChannelByTitle(state, { title: channelTitle });
+            }
+
+        Basically, it finds the first (truthy) value in the category list and that's assumed to be the
+        channel title. And then the channel details (e.g. the URL) are looked up by title (!) (not by any
+        kind of ID) via an additional API call.
+
+        TODO: Implement the API calls giving us the channel list, so that we can do the title lookup and then figure out the channel URL
+
+        May return None of no category list could be found or no category had a label ('value').
+        """
```
```suggestion
        """
        Extract the channel title, by going through the list of categories and finding the first value of the
        first category that has a value. This is how Nebula frontend extracts this information
        """
        # TODO: Implement the API calls giving us the channel list, so that we can do the title lookup and then figure out the channel URL
```

hheimbuerger replied:
Same as above. I think this is important information, and I don't think it should be stored separately from the code.

Would you prefer this as a module docstring?

pukkandan commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+        """
+        categories = video_meta.get('categories', []) if video_meta else []
+        for category in categories:
+            if category.get('value'):   # we're intentionally not using "'value' in category" here, because the expression is supposed to be falsy for empty lists in category['value'] as well!
```
```suggestion
            if category.get('value'):
```

hheimbuerger replied:
I think this is essential. If I saw this code (`if category.get('value')`), I would immediately assume an unexperienced Python developer who doesn't have a full grasp of the best practices with using Python dicts, and has therefore written it's own 'harder to read' version of `'value' in category`.

But here, there's a specific edge case which makes this code work, and the seemlingly 'cleaner' code not work. For me, that's a classic case of necessity of a comment: it states the reason for the code to be specifically what it is.

pukkandan commented on CONTRIBUTORS:
```diff
@@ -23,3 +23,4 @@ tsukumi
 bbepis
 Pccode66
 Ashish
+hheimbuerger
```
remove this. I will add it once your code is merged. Keeping this in the PR makes it necessary to conflict resolve frequently

hheimbuerger commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+            }
+        },
+    ]
+    _WORKING = True   # FIXME: should this be set to False, to hide the tests from CI, given that the unit tests require an auth cookie of a (paid) subscription?
```
Gotcha. Just for my own benefit, could you link me to documentation what this field does? I didn't quite get it.

pukkandan replied:
`_WORKING = True` is correct. Just that it is unnecessary. It only needs to be set when the extractor is broken
https://github.com/yt-dlp/yt-dlp/blob/c4a508ab31f725545bb1faba10e21bc2a6416421/yt_dlp/extractor/common.py#L400-L401

hheimbuerger commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+                self.to_screen('Authenticating to Nebula with token from --videopassword')
+
+        if not nebula_token:
+            raise ExtractorError('Nebula requires an account with an active subscription. '
```
Cool, I knew there had to be a helper somewhere!

FYI, I'm pretty sure it isn't documented, because I seem to remember searching for a good while, before just awkwardly raising my own generic exception.

hheimbuerger commented on yt_dlp/extractor/nebula.py:
```diff
@@ -0,0 +1,320 @@
+        nebula-auth cookie. Or False, if authentication fails.
+        """
+        data = json.dumps({'email': username, 'password': password}).encode('utf8')
+        request = sanitized_Request(method='POST',
+                                    url='https://api.watchnebula.com/api/v1/auth/login/',
+                                    data=data,
+                                    headers={
+                                        'content-type': 'application/json',
+                                        # Overwrite the cookie headers, because
+                                        # submitting the 'sessionid' cookie
+                                        # always causes a 403 on auth endpoint
+                                        'cookie': ''})
+        response = self._download_json(request, fatal=False, video_id=video_id,
+                                       note='Authenticating to Nebula with supplied credentials',
+                                       errnote='Authentication failed or rejected')
+        if not response or 'key' not in response:
```
Mmmh, I don't know. I suspect I tried to work around this 403 weirdness in a different way first, that needed additional capabilities of `Request`. And then when I settled on this approach, I probably forgot to simplify fully.

pukkandan replied:
prioritize this over my other reviews since this is the only part that I cant really fix myself if needed without account details

hheimbuerger replied:
@pukkandan I believe the reason I'm using a `Request` object here was that I need to POST. There's no way to pass a `method` argument to `_download_json()`, is there?

pukkandan replied:
If `data` is give, it is automatically considered `POST`

hheimbuerger replied:
Of course. And I think I figured that out once, but had already forgotten again.

This makes it a lot more readable. ✅

pukkandan commented:
@hheimbuerger It would be easier for me to keep track of any changes you are making without having to re-review the whole thing if you didn't force-push

hheimbuerger commented:
> @hheimbuerger It would be easier for me to keep track of any changes you are making without having to re-review the whole thing if you didn't force-push

I force-rebased to master. No changes yet, I'll commit those later.
But yeah, sorry, didn't quite think this through and first review I'm involved in on GitHub. Will regular-merge master next time.

pukkandan commented:
@hheimbuerger rebasing is fine (though merging/rebasing is actually unnecessary for PRs as long as there are no conflicts). I thought you made some changes.

hheimbuerger commented:
@pukkandan No changes yet. I just rebased to upstream master to make sure nothing broke in the extractor due to other changes made since. (E.g. to the Zype extractor I'm relying on.)

I will soon submit all my review changes as one commit, unless you prefer them to be in separate commits of individual changes.

hheimbuerger commented:
@pukkandan Implemented all of the suggested code changes, minus the docstring ones so far. I have to come back to those with a fresh mind on another day.

Thank you for all the suggestions! Great stuff in there, it's all so much cleaner and simpler already.

pukkandan commented:
@hheimbuerger Since only changes to comments are left now and you seem to be busy, do you mind if I make the necessary changes myself and merge?

hheimbuerger commented:
@pukkandan I moved all implementation notes and reasoning into a module docstring. Would that be an acceptable way to preserve them? I believe there's important information there for future maintenance, that I wouldn't want to store 'off-source'. I even cleaned it up some and added an important section.

All method docstrings are now down to 1-2 sentences.

Please let me know if this is a compromise you could live with.

pukkandan commented:
> Please let me know if this is a compromise you could live with.

The issue is not whether or not I could "live with" it. While I personally disagree with having so many comments, I do understand your desire to have everything documented through comments. However, this is not how this project (or youtube-dl for that matter) is maintained. You can see the other extractors yourself. In fact, you have probably written more comments than all the extractors combined!

>  I moved all implementation notes and reasoning into a module docstring

This only makes it worse. One of he many issues in depending on comments to make your code understandable is that when someone edits the code, they will most probably not edit the related comments. This issue only increases the farther away the code and its comments are.

> Would that be an acceptable way to preserve them? I believe there's important information there for future maintenance, that I wouldn't want to store 'off-source'.

I see very little actually usefull information that is not already obvious from the code.

Consider this example

```py
    def _extract_channel(self, video_meta):
        """
        Extract the channel title, by going through the list of categories and finding the first value of the
        first category that has a value.
        I know this look like a terrible approach. But actually, it's just reproducing the behavior of the
        React code the Nebula frontend uses (as of 2020-04-07):
            let channel;
            if (video && video.categories && video.categories.length) {
                const channelTitle = video.categories.map((category) => (category.value[0]))
                                                     .filter((title) => (!!title))[0];
                channel = getChannelByTitle(state, { title: channelTitle });
            }
        Basically, it finds the first (truthy) value in the category list and that's assumed to be the
        channel title. And then the channel details (e.g. the URL) are looked up by title (!) (not by any
        kind of ID) via an additional API call.
        TODO: Implement the API calls giving us the channel list, so that we can do the title lookup and then figure out the channel URL
        May return None of no category list could be found or no category had a label ('value').
        """
        categories = video_meta.get('categories', []) if video_meta else []
        for category in categories:
            if category.get('value'):   # we're intentionally not using "'value' in category" here, because the expression is supposed to be falsy for empty lists in category['value'] as well!
                return category['value'][0]
```
This has the code:
```py
        categories = video_meta.get('categories', []) if video_meta else []
        for category in categories:
            if category.get('value'):
                return category['value'][0]
```
an explaination:
```
        Extract the channel title, by going through the list of categories and finding the first value of the
        first category that has a value.
```
another reference code that does the same thing:
```
        I know this look like a terrible approach. But actually, it's just reproducing the behavior of the
        React code the Nebula frontend uses (as of 2020-04-07):
            let channel;
            if (video && video.categories && video.categories.length) {
                const channelTitle = video.categories.map((category) => (category.value[0]))
                                                     .filter((title) => (!!title))[0];
                channel = getChannelByTitle(state, { title: channelTitle });
            }
```
and again, an explaination for that reference code!!!
```
        Basically, it finds the first (truthy) value in the category list and that's assumed to be the
        channel title. And then the channel details (e.g. the URL) are looked up by title (!) (not by any
        kind of ID) via an additional API call.
```

Someone going through this has to read and understand all of these only to realize that everything here except the code is superfluous. That is not even getting started on the difficulty in maintaining this comments up-to-date with the code as changes are made.

On the other hand, If I were shown the code, simply without any comments at all, my only questions would be (1) what is the function getting? it say `extract_channel` - does it give a channel data object, title or something else. (2) What does categories have to do with channel

So, I would rewrite this to:
```py
    def _extract_channel_title(self, video_meta):
        # TODO: Implement the API calls giving us the channel list, so that we can do the title lookup and then figure out the channel URL
        categories = video_meta.get('categories', []) if video_meta else []
        # the channel name is the value of the first category
        for category in categories:
            if category.get('value'):
                return category['value'][0]
```

this is far more readable, maintainable and inline with how the rest of the code is written

---

I hope you now understand my reasoning behind asking for minimal comments, or atleast is willing to oblige it. You don't even need to do anything more. I can clean up the comments and merge this; and last week, I had even asked your opinion on me doing so

hheimbuerger commented:
@pukkandan

> this is not how this project (or youtube-dl for that matter) is maintained. You can see the other extractors yourself. In fact, you have probably written more comments than all the extractors combined!

That is not a very convincing argument to me. Gotta start somewhere.

> > I moved all implementation notes and reasoning into a module docstring
> 
> This only makes it worse. One of he many issues in depending on comments to make your code understandable is that when someone edits the code, they will most probably not edit the related comments. This issue only increases the farther away the code and its comments are.

I wholeheartedly agree! The reasoning should always be stored as close as possible to the corresponding code.

> I hope you now understand my reasoning behind asking for minimal comments, or atleast is willing to oblige it. You don't even need to do anything more. I can clean up the comments and merge this; and last week, I had even asked your opinion on me doing so

And clearly, I had read your message and wanted to make another attempt at finding a compromise between my own standards of quality and the project's. 😄

But I oblige now, you may go ahead and clean up the comments any way you want.

hheimbuerger commented:
🙌
